# Sorting, grouping, and reshaping data with DataFrames

In this lesson, you'll continue to learn how to manipulate a DataFrame by doing the following:

  - Sorting data
  - Grouping data
  - Reshaping data
    
To explore this, you'll again read in a data table consisting of the salaries and personal statistics of major league baseball players.

In [ ]:
import pandas as pd

players_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/players.csv")

First, examine your DataFrame using some of the techniques that you looked at in the previous lesson.

In [ ]:
players_df.shape

(26428, 14)

In [ ]:
players_df.describe()

birthyear    deathyear  ...        yearid        salary
count  26428.000000   492.000000  ...  26428.000000  2.642800e+04
mean    1971.389133  2008.833333  ...   2000.878727  2.085634e+06
std        9.679736     7.188803  ...      8.909314  3.455348e+06
min     1925.000000  1989.000000  ...   1985.000000  0.000000e+00
25%     1964.000000  2006.000000  ...   1994.000000  2.947020e+05
50%     1971.000000  2011.000000  ...   2001.000000  5.500000e+05
75%     1979.000000  2015.000000  ...   2009.000000  2.350000e+06
max     1995.000000  2018.000000  ...   2016.000000  3.300000e+07

[8 rows x 6 columns]

In [ ]:
players_df.columns

Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary'],
      dtype='object')

In [ ]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26428 entries, 0 to 26427
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   playerid      26428 non-null  object 
 1   birthyear     26428 non-null  int64  
 2   birthcountry  26428 non-null  object 
 3   deathyear     492 non-null    float64
 4   namefirst     26428 non-null  object 
 5   namelast      26428 non-null  object 
 6   weight        26428 non-null  int64  
 7   height        26428 non-null  int64  
 8   bats          26428 non-null  object 
 9   throws        26428 non-null  object 
 10  yearid        26428 non-null  int64  
 11  teamid        26428 non-null  object 
 12  lgid          26428 non-null  object 
 13  salary        26428 non-null  int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 2.8+ MB


In [ ]:
players_df

playerid  birthyear birthcountry  ...  teamid lgid    salary
0      barkele01       1955          USA  ...     ATL   NL    870000
1      bedrost01       1957          USA  ...     ATL   NL    550000
2      benedbr01       1955          USA  ...     ATL   NL    545000
3       campri01       1953          USA  ...     ATL   NL    633333
4      ceronri01       1954          USA  ...     ATL   NL    625000
...          ...        ...          ...  ...     ...  ...       ...
26423  strasst01       1988          USA  ...     WAS   NL  10400000
26424  taylomi02       1991          USA  ...     WAS   NL    524000
26425  treinbl01       1988          USA  ...     WAS   NL    524900
26426  werthja01       1979          USA  ...     WAS   NL  21733615
26427  zimmery01       1984          USA  ...     WAS   NL  14000000

[26428 rows x 14 columns]

## Sorting

To sort a DataFrame, use `sort_values()`. It's possible to sort by the values in one or more columns, either in ascending order or descending order. The `sort_values()` method takes several parameters:

 - `by`: Either the name of a single column or a list of names of columns
 - `axis`: Either `0` to sort rows or `1` to sort columns; defaults to `0`
 - `ascending`: `True` (or `False` for `descending`); defaults to `True`
 - `inplace`: `True` to sort in place and modify the DataFrame, and `False` to create a new DataFrame; defaults to `False`
 - `na_position`: Where to put `NaN` values, either `first` or `last`; defaults to `last`

Using what you learned in the previous lesson, again select all players that played in the team CLE in 2015.

In [ ]:
# Create a Series of booleans for the row selection
cle_options = (players_df['teamid'] == 'CLE') & (players_df['yearid'] == 2015)
cle_2015 = players_df.iloc[cle_options.values, [0, 1, 4, 5, 6, 7, 8, 9, 13]]
cle_2015

playerid  birthyear namefirst  ... bats  throws    salary
24953  adamsau01       1986    Austin  ...    R       R    507700
24954  allenco01       1988      Cody  ...    R       R    547100
24955  atchisc01       1976     Scott  ...    R       R    900000
24956  avilemi01       1981      Mike  ...    R       R   3500000
24957  bauertr01       1991    Trevor  ...    R       R   1940000
24958  bournmi01       1982   Michael  ...    L       R  13500000
24959  brantmi02       1987   Michael  ...    L       L   5875000
24960  carraca01       1987    Carlos  ...    R       R   2337500
24961  chiselo01       1988    Lonnie  ...    L       R   2250000
24962  crockky01       1991      Kyle  ...    L       L    510900
24963  floydga01       1983     Gavin  ...    R       R   4000000
24964  gomesya01       1987       Yan  ...    R       R   1000000
24965  hagadni01       1986      Nick  ...    L       L    517100
24966  housetj01       1989     T. J.  ...    R       L    511900
24967  kipnija01       1987     Jason  ...    L       R   4166666
24968  klubeco01       1986     Corey  ...    R       R   1200000
24969  mcallza01       1987      Zach  ...    R       R    520400
24970   mossbr01       1983   Brandon  ...    L       R   6500000
24971  murphda07       1981     David  ...    L       L   6000000
24972  perezro02       1988   Roberto  ...    R       R    508600
24973  raburry01       1981      Ryan  ...    R       R   2500000
24974  ramirjo01       1992      Jose  ...    B       R    511300
24975  rzepcma01       1985      Marc  ...    L       L   2400000
24976  santaca01       1986    Carlos  ...    B       R   6000000
24977   shawbr01       1987     Bryan  ...    B       R   1550000
24978  swarzan01       1985   Anthony  ...    R       R    900000
24979  swishni01       1980      Nick  ...    B       L  15000000
24980  tomlijo01       1984      Josh  ...    R       R   1500000
24981  walteza01       1989      Zach  ...    B       R    509600

[29 rows x 9 columns]

In [ ]:
# Sort the players by weight from lowest value to highest value (ascending)
# Accept all other defaults
weight_asc = cle_2015.sort_values(by='weight')
weight_asc

playerid  birthyear namefirst  ... bats  throws    salary
24974  ramirjo01       1992      Jose  ...    B       R    511300
24962  crockky01       1991      Kyle  ...    L       L    510900
24973  raburry01       1981      Ryan  ...    R       R   2500000
24957  bauertr01       1991    Trevor  ...    R       R   1940000
24958  bournmi01       1982   Michael  ...    L       R  13500000
24980  tomlijo01       1984      Josh  ...    R       R   1500000
24961  chiselo01       1988    Lonnie  ...    L       R   2250000
24967  kipnija01       1987     Jason  ...    L       R   4166666
24979  swishni01       1980      Nick  ...    B       L  15000000
24953  adamsau01       1986    Austin  ...    R       R    507700
24959  brantmi02       1987   Michael  ...    L       L   5875000
24955  atchisc01       1976     Scott  ...    R       R    900000
24966  housetj01       1989     T. J.  ...    R       L    511900
24956  avilemi01       1981      Mike  ...    R       R   3500000
24981  walteza01       1989      Zach  ...    B       R    509600
24954  allenco01       1988      Cody  ...    R       R    547100
24970   mossbr01       1983   Brandon  ...    L       R   6500000
24971  murphda07       1981     David  ...    L       L   6000000
24976  santaca01       1986    Carlos  ...    B       R   6000000
24960  carraca01       1987    Carlos  ...    R       R   2337500
24964  gomesya01       1987       Yan  ...    R       R   1000000
24968  klubeco01       1986     Corey  ...    R       R   1200000
24978  swarzan01       1985   Anthony  ...    R       R    900000
24977   shawbr01       1987     Bryan  ...    B       R   1550000
24972  perezro02       1988   Roberto  ...    R       R    508600
24975  rzepcma01       1985      Marc  ...    L       L   2400000
24965  hagadni01       1986      Nick  ...    L       L    517100
24969  mcallza01       1987      Zach  ...    R       R    520400
24963  floydga01       1983     Gavin  ...    R       R   4000000

[29 rows x 9 columns]

Suppose that you want to sort by weight and then by height. Look at the sorted DataFrame above. Notice that sometimes players with the same weight may have their heights in the wrong order. To fix this, you can provide a list of columns to sort by.

In [ ]:
weight_height_asc = cle_2015.sort_values(by=['weight', 'height'])
weight_height_asc

playerid  birthyear namefirst  ... bats  throws    salary
24974  ramirjo01       1992      Jose  ...    B       R    511300
24962  crockky01       1991      Kyle  ...    L       L    510900
24973  raburry01       1981      Ryan  ...    R       R   2500000
24958  bournmi01       1982   Michael  ...    L       R  13500000
24957  bauertr01       1991    Trevor  ...    R       R   1940000
24980  tomlijo01       1984      Josh  ...    R       R   1500000
24961  chiselo01       1988    Lonnie  ...    L       R   2250000
24967  kipnija01       1987     Jason  ...    L       R   4166666
24979  swishni01       1980      Nick  ...    B       L  15000000
24953  adamsau01       1986    Austin  ...    R       R    507700
24955  atchisc01       1976     Scott  ...    R       R    900000
24959  brantmi02       1987   Michael  ...    L       L   5875000
24956  avilemi01       1981      Mike  ...    R       R   3500000
24966  housetj01       1989     T. J.  ...    R       L    511900
24976  santaca01       1986    Carlos  ...    B       R   6000000
24954  allenco01       1988      Cody  ...    R       R    547100
24970   mossbr01       1983   Brandon  ...    L       R   6500000
24981  walteza01       1989      Zach  ...    B       R    509600
24971  murphda07       1981     David  ...    L       L   6000000
24960  carraca01       1987    Carlos  ...    R       R   2337500
24964  gomesya01       1987       Yan  ...    R       R   1000000
24968  klubeco01       1986     Corey  ...    R       R   1200000
24978  swarzan01       1985   Anthony  ...    R       R    900000
24972  perezro02       1988   Roberto  ...    R       R    508600
24977   shawbr01       1987     Bryan  ...    B       R   1550000
24975  rzepcma01       1985      Marc  ...    L       L   2400000
24965  hagadni01       1986      Nick  ...    L       L    517100
24969  mcallza01       1987      Zach  ...    R       R    520400
24963  floydga01       1983     Gavin  ...    R       R   4000000

[29 rows x 9 columns]

Suppose that you wanted to study the salaries of all players across all teams that played in the five-year period from 2011 to 2015. First, you can restrict the DataFrame to only the rows involving those years.

In [ ]:
years = (players_df['yearid'] >= 2011) & (players_df['yearid'] <= 2015)
five_years = players_df.iloc[years.values, [4, 5, 10, 13]]
five_years.head()

namefirst    namelast  yearid   salary
21454     Henry      Blanco    2011  1000000
21455    Willie  Bloomquist    2011   900000
21456     Geoff        Blum    2011  1350000
21457   Russell     Branyan    2011  1000000
21458       Sam       Demel    2011   417000

Next, create a single column with each player's full name. This way, you can use that column in the next step.

In [ ]:
five_years = five_years.assign(fullname = five_years.namefirst + ' ' + five_years.namelast)
five_years.head()

namefirst    namelast  yearid   salary           fullname
21454     Henry      Blanco    2011  1000000       Henry Blanco
21455    Willie  Bloomquist    2011   900000  Willie Bloomquist
21456     Geoff        Blum    2011  1350000         Geoff Blum
21457   Russell     Branyan    2011  1000000    Russell Branyan
21458       Sam       Demel    2011   417000          Sam Demel

At this point, the DataFrame contains the data that you want, but it's not in a format that is easy to make sense of. There is a row for each year that a player played on a team. So each player appears multiple times in the DataFrame. For example, you can look for Henry Blanco:

In [ ]:
five_years.iloc[(five_years['fullname'] == 'Henry Blanco').values]

namefirst namelast  yearid   salary      fullname
21454     Henry   Blanco    2011  1000000  Henry Blanco
22293     Henry   Blanco    2012  1250000  Henry Blanco
23902     Henry   Blanco    2013   750000  Henry Blanco

Henry Blanco played in 2011, 2012, and 2013. You can see his salary over those years too. But you want to be able to see this data for all players. It may be possible to sort by the player name, but then comparing salaries for a single year is difficult. What would help is a pivot table—one that's very similar to the pivot tables that you created in Excel. A pivot table reshapes the DataFrame from a long format to a wide format. You could create a table with the players' names in a single column and each year as a column. Each year column would then contain the salary for that player for that year.

## Grouping 

The `groupby()` method allows you to group records by one or more categorical variables. Typically, you will want to chain some data manipulation method after `groupby()` to summarize the data. 

For example, imagine that you wanted to calculate the average for your statistics, grouped by each year. First, group the DataFrame by `yearid`, then calculate the averages with `mean()`:

In [ ]:
players_df.groupby(['yearid']).mean()

birthyear    deathyear      weight     height        salary
yearid                                                               
1985    1954.670909  2007.176471  187.078182  73.298182  4.762994e+05
1986    1956.552846  2007.657895  187.535230  73.319783  4.171470e+05
1987    1957.757576  2007.031250  187.397129  73.282297  4.347295e+05
1988    1958.800905  2007.363636  187.570136  73.352941  4.531711e+05
1989    1959.907173  2006.818182  188.007032  73.378340  5.063231e+05
1990    1961.440600  2007.944444  187.876586  73.337947  5.119737e+05
1991    1961.823358  2008.000000  188.778102  73.426277  8.949612e+05
1992    1962.712614  2008.419355  188.873862  73.379714  1.047521e+06
1993    1964.230769  2008.392857  189.726977  73.401950  9.769666e+05
1994    1965.047511  2010.000000  189.771493  73.450226  1.049589e+06
1995    1966.190669  2008.950000  190.334686  73.356998  9.649791e+05
1996    1966.933405  2009.882353  191.357680  73.440387  1.027909e+06
1997    1967.793514  2010.000000  192.011892  73.430270  1.218687e+06
1998    1968.765531  2009.000000  193.331663  73.490982  1.280845e+06
1999    1969.745527  2009.928571  194.197813  73.492048  1.485317e+06
2000    1970.144737  2010.100000  194.863636  73.435407  1.992985e+06
2001    1971.012791  2009.000000  195.859302  73.476744  2.279841e+06
2002    1971.994090  2010.888889  197.478723  73.423168  2.392527e+06
2003    1973.010883  2010.000000  199.470375  73.465538  2.573473e+06
2004    1973.845969  2010.000000  200.992780  73.471721  2.491776e+06
2005    1974.767750  2011.375000  202.623345  73.546330  2.633831e+06
2006    1975.859585  2011.142857  205.477411  73.548230  2.834521e+06
2007    1976.994062  2011.000000  207.755344  73.614014  2.941436e+06
2008    1978.273364  2015.200000  209.542056  73.636682  3.136517e+06
2009    1979.195572  2013.750000  211.409594  73.697417  3.277647e+06
2010    1980.243373  2016.333333  212.522892  73.697590  3.278747e+06
2011    1981.383790  2016.000000  214.241955  73.754470  3.318838e+06
2012    1982.429245  2016.000000  213.938679  73.737028  3.458421e+06
2013    1983.611043  2016.000000  213.623313  73.730061  3.723344e+06
2014    1984.573566  2016.750000  213.018703  73.627182  3.980446e+06
2015    1985.434517  2016.500000  212.720930  73.649939  4.301276e+06
2016    1986.712778  2016.500000  212.577960  73.790152  4.396410e+06

For another example, you could calculate descriptive statistics for all records, grouped by batting and throwing arms. Try indexing the results to retrieve just the statistics for *salary*.

In [ ]:
players_df.groupby(['bats','throws'])['salary'].describe()

count          mean  ...        75%         max
bats throws                         ...                       
B    L         242.0  2.157601e+06  ...  2313750.0  15000000.0
     R        2330.0  2.021790e+06  ...  2300000.0  23125000.0
L    L        4691.0  2.170384e+06  ...  2495833.5  33000000.0
     R        2797.0  2.231821e+06  ...  2350000.0  24000000.0
R    L         726.0  1.653911e+06  ...  1975000.0  16000000.0
     R       15642.0  2.062512e+06  ...  2350000.0  33000000.0

[6 rows x 8 columns]

## Reshaping the DataFrame

### The `pivot()` function

The data that you have consists of a list of players and their salaries for the years that they played. To create a pivot table, you need to specify the index; that is, specify which column will be used to identify each row of the DataFrame. Then, specify which column values are going to become the column names of the table. Finally, specify which column will provide the values that go in the table cells.

One important consideration: the index column and column names must be unique in the DataFrame. If a player name and year are duplicated—maybe the player played for two different teams in the same year—you will get an error when trying to create the pivot table. There are many different ways that you can deal with duplicates in the data, and careful thought should be given to this problem. As a quick and simple solution here, you can simply drop all duplicates.

In [ ]:
# First, get rid of duplicates that may cause problems
five_years = five_years.drop_duplicates(['fullname', 'yearid'])

# Then create the pivot table
five_years_pivot = five_years.pivot(index='fullname', columns='yearid', values='salary')

print(five_years_pivot)

yearid                  2011        2012        2013        2014        2015
fullname                                                                    
A. J. Burnett     16500000.0  16500000.0  16500000.0  15000000.0   8500000.0
A. J. Ellis         421000.0    490000.0   2000000.0   3550000.0   4250000.0
A. J. Griffin            NaN         NaN    492500.0         NaN         NaN
A. J. Pierzynski   2000000.0   6000000.0   7500000.0   8250000.0   2000000.0
A. J. Pollock            NaN         NaN    491000.0    507000.0    519500.0
...                      ...         ...         ...         ...         ...
Zach Walters             NaN         NaN         NaN         NaN    509600.0
Zach Wheeler             NaN         NaN         NaN    512375.0         NaN
Zack Cozart              NaN    480000.0    497500.0    600000.0   2350000.0
Zack Greinke      13500000.0  13500000.0  21000000.0  26000000.0  25000000.0
Zoilo Almonte            NaN         NaN         NaN    511300.0         NaN

You may notice that the DataFrame now has *yearid* placed on top of *fullname*. This arrangement is known as *hierarchical index* or *multilevel index*. For more on advanced indexing, check the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html).

To pivot the data so that it appears in a more familiar tabular format, chain the `reset_index()` method after you `pivot()` the DataFrame:

In [ ]:
# Then create the pivot table
five_years_pivot = five_years.pivot(index='fullname', columns='yearid', values='salary').reset_index()

print(five_years_pivot)

yearid          fullname        2011  ...        2014        2015
0          A. J. Burnett  16500000.0  ...  15000000.0   8500000.0
1            A. J. Ellis    421000.0  ...   3550000.0   4250000.0
2          A. J. Griffin         NaN  ...         NaN         NaN
3       A. J. Pierzynski   2000000.0  ...   8250000.0   2000000.0
4          A. J. Pollock         NaN  ...    507000.0    519500.0
...                  ...         ...  ...         ...         ...
1517        Zach Walters         NaN  ...         NaN    509600.0
1518        Zach Wheeler         NaN  ...    512375.0         NaN
1519         Zack Cozart         NaN  ...    600000.0   2350000.0
1520        Zack Greinke  13500000.0  ...  26000000.0  25000000.0
1521       Zoilo Almonte         NaN  ...    511300.0         NaN

[1522 rows x 6 columns]


### The `melt()` function

With the `melt()` function, you can reshape a DataFrame. 

This function takes many arguments, following this form:

`pandas.melt(frame, id_vars=None, value_vars=None, var_name=None, value_name='value', col_level=None)`


Here's what the first five arguments do, according to [pandas documentation](https://www.geeksforgeeks.org/python-pandas-melt/):


| Argument     | What it does                                                                               |
| ------------ | ------------------------------------------------------------------------------------------ |
| `frame`    | DataFrame                                                    |
| `id_vars`    | Column or columns to use as identifier columns                                                    |
| `value_vars` | Column or columns to unpivot. If not specified, uses all columns that are not set as `id_vars`.    |
| `var_name`   | Name to use for the `variable` column. If none, it uses `frame.columns.name` or `variable`. |
| `value_name` | Name to use for the `value` column.                                                        |
| `col_level`  | If columns are a MultiIndex, then use this level to melt.                                   |

Using `melt()`, how can you unpivot the *five_years_pivot* table from above? Rather than having five columns for `2011`-`2015` indicating salary information, you just want one column called *year* and another called *salary*. 

That means that *fullname* is your `id_vars`, and the rest of the columns are your `value_vars`. You'll set `var_name` to *year* and `value_name` to *salary*, as previously specified. 



In [ ]:
pd.melt(frame=five_years_pivot, id_vars=['fullname'],var_name='year',value_name='salary')

fullname  year      salary
0        A. J. Burnett  2011  16500000.0
1          A. J. Ellis  2011    421000.0
2        A. J. Griffin  2011         NaN
3     A. J. Pierzynski  2011   2000000.0
4        A. J. Pollock  2011         NaN
...                ...   ...         ...
7605      Zach Walters  2015    509600.0
7606      Zach Wheeler  2015         NaN
7607       Zack Cozart  2015   2350000.0
7608      Zack Greinke  2015  25000000.0
7609     Zoilo Almonte  2015         NaN

[7610 rows x 3 columns]

## Merging DataFrames with the `merge()` function

Joining two or more tables is one of the most common data preparation tasks, and there are many ways to do it. Your understanding of joins in SQL will give you a solid grasp of how joins are conducted in pandas, because pandas joins are modeled off of SQL joins. 

The above DataFrame consisting of player information and salaries is actually the result of joining two individual DataFrames. You can reproduce this DataFrame using `merge()`. For now, call the `merge()` function and pass in the two DataFrames, assigning the results to a new DataFrame.

Pandas will make a best guess on how to merge these DataFrames; if any two columns are named the same between them, that will serve as the basis for the join. In this case, the *playerID* field is shared between the two tables, so pandas will join based on that connection.  





In [ ]:
salaries = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/salaries.csv")
people = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/people.csv")

print(salaries.columns)
print(people.columns)

Index(['yearid', 'teamid', 'lgid', 'playerid', 'salary'], dtype='object')
Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws'],
      dtype='object')


In [ ]:
merged = pd.merge(salaries, people)

print(merged.columns)
print(merged.shape)

Index(['yearid', 'teamid', 'lgid', 'playerid', 'salary', 'birthyear',
       'birthcountry', 'deathyear', 'namefirst', 'namelast', 'weight',
       'height', 'bats', 'throws'],
      dtype='object')
(26428, 14)


Indeed, this is the same information as contained in the original DataFrame, *players_df*.

In [ ]:
players_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/players.csv")

print(players_df.columns)
print(players_df.shape)

Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary'],
      dtype='object')
(26428, 14)


Should you need to perform another type of join, such as one when columns do not have the same name, you can check the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html). As mentioned earlier, your understanding of the logic behind how joins are conducted in SQL will help you grasp how to use the `merge()` function in pandas.

## Case study 1

Your manager would like to study the distribution of right-handed batters to right-handed throwers. They want to know how many right-handed batters are also right-handed throwers, and they are also interested in other combinations. You look at the original data and realize that you will have to perform some sort of count over all the rows where the value for bats is `right-handed` and the value for throws is `right-handed`, and you'll do something similar for other combinations.

The data contains duplicate rows for each player, because a row actually represents a single year per player. You do not need all of that information, so you could start by dropping the duplicate players.

In [ ]:
players_no_duplicates = players_df.drop_duplicates(['playerid'])
players_no_duplicates.shape

(5149, 14)

That leaves 5,149 players. Your manager wants a table of data that shows the counts of each combination of bats and throws. This is an ideal scenario for a *crosstab*. To create a crosstab, simply provide the two columns of data that you wish to cross-tabulate, and pandas will do the rest for you.

In [ ]:
pd.crosstab(players_no_duplicates['bats'], players_no_duplicates['throws'])

throws    L     R
bats             
B        44   400
L       918   520
R       152  3115

## Case study 2

Your company got some data about the fuel use of a number of vehicles in a relational database named *fueleconomy*. This database is on the same server as the previous database. You would like to answer some questions about this data, such as the following:

1. How many vehicles use each type of fuel by year?
2. What are the top-performing six-cylinder vehicles in 2003, as measured by highway mileage?


First, connect to the database and get the data into a DataFrame.


In [ ]:
vehicles_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/vehicles.csv")

Next, look at some features of the data that you got. For example, how many rows are there? What are the data types?

In [ ]:
vehicles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33442 entries, 0 to 33441
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      33442 non-null  int64  
 1   make    33442 non-null  object 
 2   model   33442 non-null  object 
 3   year    33442 non-null  int64  
 4   class   33442 non-null  object 
 5   trans   33434 non-null  object 
 6   drive   33442 non-null  object 
 7   cyl     33384 non-null  float64
 8   displ   33385 non-null  float64
 9   fuel    33442 non-null  object 
 10  hwy     33442 non-null  int64  
 11  cty     33442 non-null  int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 3.1+ MB


Next, you can check how many years of data are contained in the DataFrame. That is, how many unique years are there?

In [ ]:
vehicles_df['year'].value_counts()

1985    1701
1987    1247
2014    1214
1986    1210
2008    1186
2013    1184
2009    1184
2005    1166
1989    1153
2012    1153
1991    1132
1988    1130
2011    1130
2007    1126
2004    1122
1992    1121
2010    1109
2006    1104
1993    1093
1990    1078
2003    1044
1994     982
2002     974
1995     967
2001     909
1999     849
2000     839
1998     811
1984     784
1996     773
1997     762
2015     205
Name: year, dtype: int64

Looks like you have data from 1985 to 2015. How many different types of fuel are there?

In [ ]:
vehicles_df['fuel'].value_counts()

Regular                        22622
Premium                         8617
Gasoline or E85                 1043
Diesel                           874
Premium or E85                    88
CNG                               58
Electricity                       55
Midgrade                          43
Gasoline or natural gas           18
Gasoline or propane                8
Regular Gas and Electricity        8
Premium Gas or Electricity         7
Premium and Electricity            1
Name: fuel, dtype: int64

Then do a crosstab between these two columns and see what you learn.

In [ ]:
pd.crosstab(vehicles_df['year'], vehicles_df['fuel'])

fuel  CNG  Diesel  ...  Regular  Regular Gas and Electricity
year               ...                                      
1984    0     105  ...      679                            0
1985    0     158  ...     1442                            0
1986    0      71  ...     1064                            0
1987    0      56  ...     1102                            0
1988    0      31  ...      995                            0
1989    0      33  ...      978                            0
1990    0      29  ...      921                            0
1991    0      39  ...      895                            0
1992    0      23  ...      877                            0
1993    1      22  ...      861                            0
1994    1      31  ...      746                            0
1995    1      32  ...      733                            0
1996    1      13  ...      612                            0
1997    1      11  ...      581                            0
1998    3      14  ...      637                            0
1999    5      10  ...      656                            0
2000    5       6  ...      576                            0
2001    7       6  ...      598                            0
2002    5       8  ...      628                            0
2003    4       9  ...      629                            0
2004    8      11  ...      635                            0
2005    5      14  ...      639                            0
2006    1       9  ...      621                            0
2007    1       7  ...      595                            0
2008    1       7  ...      591                            0
2009    1      12  ...      665                            0
2010    1      14  ...      590                            0
2011    2      16  ...      539                            0
2012    2      17  ...      503                            1
2013    1      22  ...      487                            3
2014    1      35  ...      490                            4
2015    0       3  ...       57                            0

[32 rows x 13 columns]

That gives you a useful look at the way that different fuels have been used over the years. In an upcoming lesson, you will learn how to make charts out of this type of data. 

The second question will require a few steps. First, select only the six-cylinder vehicles from 2003.

In [ ]:
bools = (vehicles_df['cyl'] == 6) & (vehicles_df['year'] == 2003)
six_cyl_df = vehicles_df.iloc[bools.values]
six_cyl_df.head()

id   make    model  year  ... displ     fuel hwy  cty
25   18458  Acura    3.2CL  2003  ...   3.2  Premium  26   17
26   18459  Acura    3.2CL  2003  ...   3.2  Premium  27   17
31   18629  Acura    3.2TL  2003  ...   3.2  Premium  27   17
39   18630  Acura    3.5RL  2003  ...   3.5  Premium  22   16
120  19118  Acura  MDX 4WD  2003  ...   3.5  Premium  21   15

[5 rows x 12 columns]

Next, you can either sort the entire DataFrame by highway mileage in descending order, or use the built-in `nlargest()` method to select the *n* rows with the largest values in the highway column. You can look at both solutions just to compare approaches.

In [ ]:
# Sort the DataFrame by 'hwy' in descending order
sorted_six_cyl_df = six_cyl_df.sort_values(by='hwy', ascending=False)

# Select the first 10 rows
sorted_six_cyl_df.head(10)

id        make        model  year  ... displ     fuel hwy  cty
5450   18731   Chevrolet       Impala  2003  ...   3.4  Regular  29   19
5873   18646   Chevrolet  Monte Carlo  2003  ...   3.4  Regular  29   19
32123  18619  Volkswagen        Jetta  2003  ...   2.8  Premium  28   19
1403   18469         BMW         325i  2003  ...   2.5  Premium  27   18
1637   18473         BMW         330i  2003  ...   3.0  Premium  27   18
3056   18641       Buick        Regal  2003  ...   3.8  Regular  27   17
2984   18727       Buick  Park Avenue  2003  ...   3.8  Regular  27   17
15922  18658       Honda       Accord  2003  ...   3.0  Regular  27   18
15921  18657       Honda       Accord  2003  ...   3.0  Regular  27   19
1609   18400         BMW        330ci  2003  ...   3.0  Premium  27   18

[10 rows x 12 columns]

That did the trick. You have the top ten vehicles by highway mileage in 2003. Notice that there are several vehicles with the same mileage values. Now, see how  `nlargest` may work to do the same job.

`nlargest` is really just a shortcut. You need to provide the following:

 - `n`: The number of rows to return
 - `columns`: The columns to sort by
 - `keep`: One of `first`, `last`, or `all`. This affects how you deal with duplicate values in the sort column, and it defaults to `first`.

In [ ]:
# Select the 10 rows with largest values in the 'hwy' column
six_cyl_df.nlargest(10, columns='hwy')

id        make             model  year  ... displ     fuel hwy  cty
5450   18731   Chevrolet            Impala  2003  ...   3.4  Regular  29   19
5873   18646   Chevrolet       Monte Carlo  2003  ...   3.4  Regular  29   19
32123  18619  Volkswagen             Jetta  2003  ...   2.8  Premium  28   19
26     18459       Acura             3.2CL  2003  ...   3.2  Premium  27   17
31     18629       Acura             3.2TL  2003  ...   3.2  Premium  27   17
1376   18398         BMW             325ci  2003  ...   2.5  Premium  27   18
1403   18469         BMW              325i  2003  ...   2.5  Premium  27   18
1430   18761         BMW  325i Sport Wagon  2003  ...   2.5  Premium  27   18
1609   18400         BMW             330ci  2003  ...   3.0  Premium  27   18
1637   18473         BMW              330i  2003  ...   3.0  Premium  27   18

[10 rows x 12 columns]

The exact rows that are returned may not exactly match because of the way that `nlargest` deals with duplicates. By default, it returns the first rows in the set of duplicates. Experiment with this feature by trying `last` and `all` as values to the `keep` parameter.